### DISCUSSION TITLE
Where to get started debugging?

### DISCUSSION THREAD

**Discussion by mechachoi** [ravinrabbid/DonCon2040#6] [General]
Alright, I've got the board all assembled and I believe it's assembled correctly. I have my drum (actually an official Namco Bandai Arcade drum, so I'm essentially trying to retrofit a real arcade setup to work with the console games). I've also connected the outputs of the piezo's to an oscilloscope and can confirm that they're working. 

I have the piezos from the arcade machine connected to the board, but I'm not seeing any "hits" or presses when I connect the board. Not sure where to start debugging here, if I should just start with changing the trigger thresholds? Or if there's an easier way to figure out what's going on?

Is there a way for me to output the ADC values from the Piezo's in this setup? Looking for any help or guidance on where to start debugging.

**Comment by ravinrabbid** [comment]
Hi,

since you have your oscilloscope already at hand, I'd probably start with probing pins 1-4 of the MCP3204 to check if any signal arrives there. This should rule out any issues with the analog part. You can also try shortly bridging Vref to one of those pins and check if this does trigger a hit. Maybe also check all Vcc pins for the correct voltages while you are at it, just to be sure.

The next step would be changing to 'Debug' mode and connect via serial terminal. This will show you the ADC readings if a hit is detected, so I recommend lowering the thresholds for this. (Mind that the values are mapped to 16bit and a peak will stay for the duration of 'Hold Time').

If the analog part is good and you still don't get any output in 'Debug' mode, I'd suspect an issue with the ADC or level-shifter.

Further, are you using the pre-built firmware or did you build it yourself?


Anyways, cool project! I'd really be interested how well this performs with an actual arcade drum.

**Reply by mechachoi** [reply]
I am using your firmware but am compiling it in VSCode with the RPi Pico as per your instructions! 

And thanks for the tips, I'll give that a try and hopefully report back success in the next few days.

**Reply by mechachoi** [reply]
One question: Is it not possible to run USB DEBUG mode without having made the DonConPad? I ask because when I try to run this mode, serial monitor only outputs "Issd1306_writel addr not acknowledged!" -- is the Pad a prerequisite to running USB DEBUG mode?

**Reply by ravinrabbid** [reply]
That could be an issue in general, I don't know how well the firmware behaves without the pad attached. From a quick test it looks like the drum triggers work while all other buttons wildly toggle on and off. And in debug mode the hit values will scroll past pretty quickly between the error messages.

I'll see if I can make some modifications so that it works properly without the pad attached. In the meantime you could try to comment line 56, all lines containing `display.` and line 154 in `main.cpp` as a quick fix.

**Reply by mechachoi** [reply]
OK-- I've commented out those lines. But I'm unfortunately still not getting any hits in the Serial Monitor, even with those lines commented out. I've been looking at the pins on the MCP3204 1-4 and not seeing anything coming up on the scope, although I do see a signal coming when I probe the lines of the drum piezos directly. When I short the MCP3204 lines 1-4 to VRef, nothing happens. I checked the 5V line into MCP3204 and it's approx 5.2 V, and the Vref on Pin 2 is about 4V. 

I'm sort of running out of ideas on how to fix this-- I wonder if something is screwed up with the analog part?

**Reply by mechachoi** [reply]
![image](https://github.com/user-attachments/assets/cd560e5c-96ee-4ccd-bf2b-2f6b2ebe5a12)

**Reply by ravinrabbid** [reply]
Hmm, that's indeed strange. I don't see any obvious problems with your board.

Since none of the analog inputs seem to yield any output, can you check the +12V and -12V rails on the TLC072 opamps? Further you try to 'trace' the signal through the circuit and see where it gets lost, basically follow the horizontal path between the '\*_IN' and '\*_ADC' labels in the schematic.

It is further weird that you can't trigger the inputs directly. Maybe try to flash the pre-built firmware from the releases page to rule out any issues with you particular build, just to be sure.

**Reply by mechachoi** [reply]
I'm getting about 13V coming out of the converter and at all the rails of the Op-Amps-- is that normal?

**Reply by mechachoi** [reply]
I'm not sure what changed, but I am definitely seeing the signals coming into those MCP3204 pins that I have attached wires. What should I do to make sure the MCP3204 is "outputting" the right signal to the U7?

**Reply by ravinrabbid** [reply]
Just for reference, I'm measuring +/-12.6V, +3.33V and +5.1V respectively. Slightly lower, but your measurements should still be well within tolerance.

U7 does the level shifting between 3.3V and 5V for the ADC's SPI communication. Pin 8 should be a constant 3.3V, the rest of the pins should mirror the signal of the opposing pin with a different voltage. Pin 5/10 should be a clock signal, Pin 2/13 should be pulled low in regular intervals. On Pin 4/11 and 3/12 you should see some data being exchanged.

**Comment by mechachoi** [comment]
I think I've gotten further than I've been before! I have it in DEBUG mode and its outputting:

`(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]
(*( 4095[########]( | 4095[########]| ) 4095[########])*) 4095[########]`

repeatedly. Not sure what this means, as I'm having a hard time parsing the USB_DEBUG code. Any ideas what might be happening?

**Reply by mechachoi** [reply]
I'm having a hard time finding where in the code this print is called... any help is appreciated!

**Reply by ravinrabbid** [reply]
The debug line is constructed in https://github.com/ravinrabbid/DonCon2040/blob/58cf7435f5e558668160dd035892bbd731bb20fc/src/utils/InputState.cpp#L370

It's meaning is as follows: 
`( (` `( |` `| )` `) )` are Left Ka, Left Don, Right Don, Right Ka respectively, a  `*` tells you if it is considered triggered or not after resolving the false hit detection.
The number is the raw value read from the ADC (0-4095), followed by a bar for visual representation. There will only be output as long as at least one pad triggers. There should be newlines after each set of readings, you might need to configure your serial terminal for Unix line endings.

Since all your ADC inputs show the maximum value constantly, I suspect there either is an issue with the inputs to the ADC, it's reference Voltage, or the SPI MISO line being high constantly.

**Reply by mechachoi** [reply]
OK, that's helpful! Thank you. (And thanks again for being so helpful and patient with my questions!)

At this point, I know that SPI MISO is not being high constantly (output shown here):

![DF15114A-0ECD-4124-A0B3-64F853942AF2_1_102_a](https://github.com/user-attachments/assets/f7e406c3-f98a-4475-9a60-33da48331341)

It looks like the 3.3V line looks OK on the 2040. I also confirmed that I'm seeing the drum signals arrive at the MCP. I see about a 1V spike for KA and a 4V spike for DON. 

How do you recommend I check the inputs to the ADC? I'm assuming it's something happening between U5 and U7, but I could be wrong.

**Reply by ravinrabbid** [reply]
Hmm, maybe check the pin pairs 5/10 4/11 4/12 and 2/13 to see if the signals are the same on both sides of U7 (with 3.3V/5V respectively) and if they look like valid SPI (clock is running, CS goes low for transfer etc.). If so and the drum signals and the reference voltage on U5 pin 13 looks correct, I'm somewhat out of ideas ... maybe the ADC is defective

**Reply by mechachoi** [reply]
What should VRef on U5 be showing?

**Reply by ravinrabbid** [reply]
VRef should be very close to 4.096V
